graph/graph_windows.go

```diff
@@ -155,6 +182,50 @@ func (graph *Graph) TarLayer(img *Image) (arch archive.Archive, err error) {
 		// We keep this functionality here so that we can still work with the VFS
 		// driver during development. VFS is not supported (and just will not work)
 		// for Windows containers.
-		return graph.driver.Diff(img.ID, img.Parent)
+		root := graph.imageRoot(img.ID)
+		mFileName := filepath.Join(root, tarDataFileName)
+		mf, err := os.Open(mFileName)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				logrus.Errorf("failed to open %q: %s", mFileName, err)
+			}
+			logrus.Debugf("[graph] TarLayer with traditional differ: %s", img.ID)
+			return graph.driver.Diff(img.ID, img.Parent)
+		}
+		pR, pW := io.Pipe()
+		// this will need to be in a goroutine, as we are returning the stream of a
+		// tar archive, but can not close the metadata reader early (when this
+		// function returns)...
+		go func() {
+			defer mf.Close()
+			// let's reassemble!
+			logrus.Debugf("[graph] TarLayer with reassembly: %s", img.ID)
+			mfz, err := gzip.NewReader(mf)
+			if err != nil {
+				pW.CloseWithError(fmt.Errorf("[graph] error with %s:  %s", mFileName, err))
+				return
+			}
+			defer mfz.Close()
+
+			// get our relative path to the container
+			fsLayer, err := graph.driver.Get(img.ID, "")
+			if err != nil {
+				pW.CloseWithError(err)
+				return
+			}
+			defer graph.driver.Put(img.ID)
+
+			metaUnpacker := storage.NewJSONUnpacker(mfz)
+			fileGetter := storage.NewPathFileGetter(fsLayer)
+			logrus.Debugf("[graph] %s is at %q", img.ID, fsLayer)
+			ots := asm.NewOutputTarStream(fileGetter, metaUnpacker)
```

Comment:
Just pipe this output with a digest verifier against the digest, if it exists. Return a read error if that fails the check.